topics/integrate-in-existing-app.md

```diff
@@ -1,242 +1,423 @@
+    ```
+
+    ```kotlin
+    //After
+    return Result.Error(RuntimeException("Error logging in", e))
+    ```
+
+2. For email validation replace the package `android.utils` with a Kotlin regular expression matching the pattern in the `LoginDataValidator` class:
+
+    ```kotlin
+    // Before
+    private fun isEmailValid(email: String) = Patterns.EMAIL_ADDRESS.matcher(email).matches()
+    ```
+
+    ```kotlin
+    // After
+    private fun isEmailValid(email: String) = emailRegex.matches(email)
+    
+    companion object {
+       private val emailRegex =
+               ("[a-zA-Z0-9\\+\\.\\_\\%\\-\\+]{1,256}" +
+                       "\\@" +
+                       "[a-zA-Z0-9][a-zA-Z0-9\\-]{0,64}" +
+                       "(" +
+                       "\\." +
+                       "[a-zA-Z0-9][a-zA-Z0-9\\-]{0,25}" +
+                       ")+").toRegex()
     }
-}
+    ```
+
+#### Connect to platform-specific APIs from the cross-platform code {initial-collapse-state="collapsed"}
+
+A universally unique identifier (UUID) for `fakeUser` in `LoginDataSource` is generated using the `java.util.UUID` class, which is not available for iOS. 
+
+```kotlin
+val fakeUser = LoggedInUser(java.util.UUID.randomUUID().toString(), "Jane Doe")
 ```
 
-Note the `displayModel` and `lastModel` mechanism, which allows a view to detach and re-attach, for example for configuration changes on Android.
+Since the Kotlin standard library doesn't provide functionality for generating UUID, you still need to use platform-specific functionality for this case. 
+Provide the `expect` declaration for the `randomUUID()` function in the shared code and its `actual` implementations for each platform – Android and iOS – in the corresponding source sets. 
+You can learn more about [connecting to platform-specific APIs](connect-to-platform-specific-apis.md).
+
+1. Remove the `java.util.UUID` from the common code:
+
+    ```kotlin
+   val fakeUser = LoggedInUser(randomUUID(), "Jane Doe") 
+   ```
+
+1. Create the `Utils.kt` file in the `shared/src/commonMain` and provide the `expect` declaration:
+
+    ```kotlin
+    package com.jetbrains.simplelogin.shared
+    
+    expect fun randomUUID(): String
+    ```
+
+2. Create the` Utils.kt` file in `shared/src/androidMain` and provide the `actual` implementation for `randomUUID()` in Android:
+
+    ```kotlin
+    package com.jetbrains.simplelogin.shared
+    
+    import java.util.*
+    actual fun randomUUID() = UUID.randomUUID().toString()
+    ```
+
+3. Create the `Utils.kt` file in `shared/src/iosMain` and provide the `actual` implementation for `randomUUID()` in iOS:
+
+    ```kotlin
+    package com.jetbrains.simplelogin.shared
+    
+    import platform.Foundation.NSUUID
+    actual fun randomUUID(): String = NSUUID().UUIDString()
+    ```     
+
+For Android and iOS, Kotlin will use different platform-specific implementations.
+
+#### Run your cross-platform application
```

Comment:
It's better to be the third-level (###) heading, I guess. Now it's almost invisible.
With instruction on how to run and what the result we got.